R/model_functions.R

```diff
@@ -192,60 +192,56 @@ run_cv <- function(focal_vs_comp, max_dist, cv_grid, prior_param = NULL, run_shu
     pull(foldID) %>%
     unique() %>%
     sort()
-  focal_trees <- vector(mode = "list", length = length(folds))
-
-  for (i in 1:length(folds)) {
-    # Define test set and "full" training set (we will remove buffer region below)
-    test <- focal_vs_comp %>%
-      filter(foldID == folds[i])
-    train_full <- focal_vs_comp %>%
-      filter(foldID != folds[i])
-
-    # If no trees in test skip, skip to next iteration in for loop
-    if (nrow(test) == 0) {
-      next
-    }
-
-    # Define sf object of boundary of test fold
-    test_fold_boundary <- cv_grid %>%
-      subset(folds == i) %>%
-      st_bbox() %>%
-      st_as_sfc()
-
-    # Remove trees in training set that are part of test set and buffer region to test set
-    train <- train_full %>%
-      st_as_sf() %>%
-      add_buffer_variable(direction = "out", size = max_dist, region = test_fold_boundary) %>%
-      filter(buffer) %>%
-      as_tibble()
-
-    if (FALSE) {
-      # Visualize test set trees + boundary
-      ggplot() +
-        geom_sf(data = train %>% sample_frac(0.01) %>% st_as_sf(), col = "black", alpha = 0.1) +
-        geom_sf(data = test_fold_boundary, col = "orange", fill = "transparent") +
-        geom_sf(data = test %>% st_as_sf(), col = "orange")
-    }
 
-    # Fit model on training data and predict on test
-    posterior_param_fold <- train %>%
-      fit_bayesian_model(prior_param = prior_param, run_shuffle = run_shuffle)
+  purrr::map_dfr(
+    folds,
+    fit_one_fold,
```

Comment:
Let me see if I understand the `.x = folds` and `.f = fit_one_fold()` arguments:
- `.x` is always what's being iterated over, amirite? In this case, the integers 1 thru n_folds.
- The first argument of `.f` has to *necessarily* be something that indexes what's being iterated over, correct? i.e. the first argument of `.f` makes the connection from the vector of all `folds` numbers in `map_dfr()` to the specific `fold` number considered in a particular instance of `fit_old_fold()` amirite?

Reply:
* `.x` is what's iterated over, yes! In this case, `.x` is indeed an integer vector, but not necessarily `1:n_folds`.

``` r
> unique(focal_vs_comp_ex$focal_ID)
[1] 2 3 4 5 7 9
```

* `.x` does have to be an argument to `.f`, though doesn't necessarily have to be the first one. In the case that it's not the first argument, you need to specify all the preceding arguments _by name_ as arguments following `.f` within `map()` so that `map()` knows where to send which arguments. Since we have the convenience of being able to define our own function internally, I put the argument being iterated over first so that the arguments being passed by name align with what would have been passed by order by default.

`fit_one_fold()`'s iterand (thing being iterated over) happens to be a number that serves as an index for a fold, but I'm not solid enough on how "index" is used in statistics to say that that's always what `.x` is.

``` r
library(tidyverse)
  
pets <-
  tibble(
    animal = c(rep("cat", 3), rep("dog", 2), rep("bunny", 2)),
    cuteness = rnorm(7, 4)
  )

pets
#> # A tibble: 7 x 2
#>   animal cuteness
#>   <chr>     <dbl>
#> 1 cat        5.64
#> 2 cat        5.15
#> 3 cat        2.13
#> 4 dog        2.66
#> 5 dog        2.50
#> 6 bunny      3.45
#> 7 bunny      3.87

filter_animals <- function(name, df) {
  df %>% filter(animal == name)
}

# run the function on two different animals
map(
  .x = c("cat", "dog"),
  filter_animals,
  pets
)
#> [[1]]
#> # A tibble: 3 x 2
#>   animal cuteness
#>   <chr>     <dbl>
#> 1 cat        5.64
#> 2 cat        5.15
#> 3 cat        2.13
#> 
#> [[2]]
#> # A tibble: 2 x 2
#>   animal cuteness
#>   <chr>     <dbl>
#> 1 dog        2.66
#> 2 dog        2.50

# the _dfr suffix just row binds the results
map_dfr(
  .x = c("cat", "dog"),
  filter_animals,
  pets
)
#> # A tibble: 5 x 2
#>   animal cuteness
#>   <chr>     <dbl>
#> 1 cat        5.64
#> 2 cat        5.15
#> 3 cat        2.13
#> 4 dog        2.66
#> 5 dog        2.50
```

<sup>Created on 2021-01-18 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

Here, each element of `.x` is an string whose value can be matched inside of a data frame column. This is what's happening inside of `run_cv` as well... the `foldID`s _are_ indices for folds, but as arguments to `fit_one_fold()`, they're just part of a condition used to subset rows.

Not sure if this is what your question was getting at—ignore if it seems I misunderstood. :-)